### PR TITLE
feat: Implement ingested_forecast_length utility and integrate with GFS (#412)

### LINKED ISSUE #412: Calculate and write `ingested_forecast_length` for GEFS, GFS, HRRR, and IFS ENS forecast datasets
We have an `ingested_forecast_length` coordinate that is currently empty. It should report the maximum lead time present in the dataset at each `init_time`. This allows users to quickly determine which initialization times are complete enough for their needs. 

This can be done in each dataset's `RegionJob` subclass's `update_template_with_results`. The existing coordinate values in the array will need to be read and then updated. A shared utility might help future integrations be bug free.

Add unit tests as well as test in the dynamical_dataset_integration_test.py for each dataset as part of the testing of a dataset `update`.

Backfill handling: backfills are parallelized and they can't/shouldn't all attempt to write to the same coordinate array. One option I see is a script that we can run after a backfill that reads data at a single point and calculates and writes this array. This can be a follow on. Perhaps there are better ways that are also simple.

### PR DESCRIPTION
**Pull request by ArkVex**
## Description
This PR implements the logic to calculate and populate the `ingested_forecast_length` coordinate for the GFS dataset, as requested in #412.

This metric helps downstream users determine the maximum available lead time for each initialization time, allowing them to filter for "complete" forecasts.

## Changes
1.  **Shared Utility:** Created `src/reformatters/common/ingest_stats.py` with a new function `update_ingested_forecast_length`.
    * Defined a `HasTimeInfo` Protocol to ensure type safety when processing coordinates.
    * Logic identifies the maximum lead time per `init_time` and updates the dataset in place.
2.  **GFS Integration:** Updated `src/reformatters/noaa/gfs/region_job.py`.
    * Overrode `update_template_with_results` to call the new utility after the standard update process.
3.  **Testing:** Added unit tests in `tests/common/test_ingest_stats.py` to verify:
    * Correct calculation of max lead times.
    * Proper handling of updates (overwriting smaller values with larger ones).
    * Handling of `pd.NaT` and empty states.

## Testing
* Added new unit tests: `tests/common/test_ingest_stats.py`.
* Verified type checking passes with `ty` (using `type: ignore` for specific pandas timedelta edge cases).

## Related Issue
Closes #412

CC @aldenks @JackKelly

### CODE REVIEW

**Comment by ArkVex** [comment]
@aldenks A GENTLE reminder

**Comment by JackKelly** [comment]
Hi @ArkVex, if the PR is ready for review then it might be worth updating the title of the PR, and fix the failing tests, and "request" a review via the "Reviewers" setting on the top right :slightly_smiling_face:.

Also, it's worth noting that Alden is a busy guy! `reformatters` evolves faster than most open source projects, but it's not uncommon for PRs to open source projects to go for weeks before being reviewed.

**Comment by ArkVex** [comment]
Ohh I apologise I was not aware of this...sure I'll fix the errors and change the pull request heading.
Thankyou @JackKelly

**Comment by aldenks** [comment]
ah yeah @ArkVex, when you're ready for a review use the "Request Review" functionality github has in the right sidebar of the PR to ask for a review from me and i'll give you one!  (i saw you marked as "draft" and the CI pipeline was failing and assumed it was a work in progress)

**Comment by ArkVex** [comment]
Hi @aldenks and @JackKelly, I've checked the linting locally and it's passing. I'm having some trouble with the GitHub permissions for the 'Reviewers' sidebar, so I'm marking this as Ready for Review now. Please let me know if the CI failures on your end still look like linting issues to you!

``` 
(reformatters) PS C:\Users\Lenovo\OneDrive\Desktop\reformatters> uv run ruff check
All checks passed! ```

File: tests/common/test_ingest_stats.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def test_update_ingested_forecast_length_simple() -> None:
+    # 1. Setup a dummy dataset
+    init_times = [
+        pd.Timestamp("2025-01-01 12:00"),
+        pd.Timestamp("2025-01-01 18:00"),
+    ]
+
+    # We use 'cast' to silence the strict type checker here
```
Comment (JackKelly) [inline tests/common/test_ingest_stats.py]: Please remove this comment :)

**Comment by JackKelly** [comment]
I could be wrong but I think you can fix the ruff errors just by removing line 2 of `test_ingest_stats.py` (i.e. remove `from typing import Any, cast`)

File: tests/common/test_ingest_stats.py (hunk 1 above)
Comment (ArkVex) [reply tests/common/test_ingest_stats.py]: 🫣

File: src/reformatters/common/ingest_stats.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    results_coords: Sequence[HasTimeInfo],
+) -> None:
+    """
+    Updates the 'ingested_forecast_length' coordinate in the template dataset.
+    """
+    if "ingested_forecast_length" not in template_ds.coords:
+        log.warning(
+            "ingested_forecast_length coordinate not found in template dataset."
+        )
+        return
```
Comment (aldenks) [inline src/reformatters/common/ingest_stats.py]:
```suggestion
    assert "ingested_forecast_length" in template_ds.coords
```

File: src/reformatters/common/ingest_stats.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+from reformatters.common.logging import get_logger
+from reformatters.common.types import Timedelta, Timestamp
+
+log = get_logger(__name__)
+
+
+# This Protocol tells the type checker: "Trust me, these objects have time info"
+class HasTimeInfo(Protocol):
+    init_time: Timestamp
+    lead_time: Timedelta
```
Comment (aldenks) [inline src/reformatters/common/ingest_stats.py]:
```suggestion
class DeterministicForecastSourceFileCoord(Protocol):
    init_time: Timestamp
    lead_time: Timedelta
```

File: src/reformatters/common/ingest_stats.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            ]
+
+            # Use .values and pd.isnull to safely check for NaT (Not a Time)
+            if pd.isnull(current_val.values) or max_lead > current_val:
+                log.info(
+                    f"Updating ingested_forecast_length for {init_time} to {max_lead}"
+                )
+                template_ds["ingested_forecast_length"].loc[
+                    {"init_time": init_time}
+                ] = max_lead
```
Comment (aldenks) [inline src/reformatters/common/ingest_stats.py]: We don't want to look at existing values because of the way we update datasets by overwriting everything in a shard, so overwriting with whatever we processed this run is correct. In practice, we make sure we're only adding to a dataset, but that happens outside of here.
```suggestion
        template_ds["ingested_forecast_length"].loc[
                {"init_time": init_time}
        ] = max_lead
```

File: src/reformatters/common/ingest_stats.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+# This Protocol tells the type checker: "Trust me, these objects have time info"
+class HasTimeInfo(Protocol):
+    init_time: Timestamp
+    lead_time: Timedelta
+
+
+def update_ingested_forecast_length(
+    template_ds: xr.Dataset,
+    results_coords: Sequence[HasTimeInfo],
```
Comment (aldenks) [inline src/reformatters/common/ingest_stats.py]: let's allow callers to pass in the `process_results` directly and handle taking the max across variable names (the str key in this Mapping are variable names) within this function, rather than needing to make all callers do the same flattening into a `Sequence[DeterministicForecastSourceFileCoord]`
```suggestion
    results_coords: Mapping[str, Sequence[DeterministicForecastSourceFileCoord]] ,
```

Then also add to the docstring the note that "The maximum processed lead time across all variables is set as the ingested_forecast_length. This can hide the nuance of a specific variable having fewer lead times processed than others."

File: src/reformatters/common/ingest_stats.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+) -> None:
```
Comment (aldenks) [inline src/reformatters/common/ingest_stats.py]:
```suggestion
) -> xr.Dataset:
```
lets have this return the modified dataset so callers would do `ds = update_ingested_forecast_length(...)`

File: src/reformatters/noaa/gfs/region_job.py [hunk 7]
```diff
@@ -150,6 +151,26 @@ def apply_data_transformations(
...
+        # process_results is { "filename": [coord1, coord2], ... }
+        all_coords = []
+        for coord_list in process_results.values():
+            all_coords.extend(coord_list)
+
+        # 3. Run our new logic
+        update_ingested_forecast_length(ds, all_coords)
+
+        # 4. Return the modified dataset (Crucial!)
+        return ds
```
Comment (aldenks) [inline src/reformatters/noaa/gfs/region_job.py]: less code is easier to understand code!
```suggestion
        ds = super().update_template_with_results(process_results)
        return update_ingested_forecast_length(ds, process_results)
```

File: tests/common/test_ingest_stats.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
+from collections.abc import Mapping
+from typing import Any, cast
```
Comment (aldenks) [inline tests/common/test_ingest_stats.py]: remove, unused

File: src/reformatters/common/ingest_stats.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    """
+    Updates the 'ingested_forecast_length' coordinate in the template dataset.
+    """
```
Comment (aldenks) [inline src/reformatters/common/ingest_stats.py:27]: Add docstring note that this updates the dataset coordinate in place

File: tests/common/test_ingest_stats.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    ).values == pd.Timedelta(hours=6)
+    assert ds["ingested_forecast_length"].sel(
+        init_time="2025-01-01 18:00"
+    ).values == pd.Timedelta(hours=48)
+
+
+def test_update_ingested_forecast_length_update_existing() -> None:
+    init_time = pd.Timestamp("2025-01-01 12:00")
+
+    # Start with 6 hours already recorded
```
Comment (aldenks) [inline tests/common/test_ingest_stats.py]: Please remove all the comments in this file except for this one. This one is helpful because it highlights the case we're testing

**Review by aldenks — CHANGES_REQUESTED** [review]
Thank you @ArkVex! This review is a little picky and i hope it doesn't scare you off.

Here's the two overarching motivations behind my review
1. Philosophy towards comments: Write code that explains itself rather than needs comments. Comments are additional code to maintain and less to maintain is better. They also make it ~2x more to read which takes time and mental space and worst of all can be out of sync with the actual code and cause more confusion. Instead, use variable and function names and organize function logic so the code itself tells you what it's doing (you've done that). Imo comments _are_ great to 1. explain something surprising or out of the ordinary (a gotcha, or where an optimization requires less clear code) 2. in select places to help explain the "why" behind something if we can't make the code say it.
1. Don't smooth over errors, rather fail early if something doesn't match expectations (or expected failure modes). If an error is silently (including a log line) "handled" its still there, it will just turn up somewhere later, far away from its source and be harder to debug.

Please also add additional test coverage:
- test that all ingested_forecast_length coordinate values that don't have init times in the process results are _not_ modified
- add a check in `tests/noaa/gfs/forecast/dynamical_dataset_test.py::tests/noaa/gfs/forecast/dynamical_dataset_test.py` that checks the coordinate value after the update step runs. This will ensure the updated template really threads through correctly and is written to the final zarr store. That test only processes a few lead times (for speed) so I'd expect the ingested_forecast_length for 2021-05-01T12:00:00 to be 3h

**Comment by ArkVex** [comment]
Thank you for the detailed feedback, @aldenks  ! I really appreciate the breakdown of the project's philosophy. I’ll refactor the utility to handle the mapping directly, clean up the comments, and add those extra test cases today.

**Comment by ArkVex** [comment]
@aldenks I feel now the PR is ready for the review...I tried my best to implement your reviews as much as possible :)

**Review by aldenks — CHANGES_REQUESTED** [review]
Hi @ArkVex, please do a closer self review of these changes.

Most of my comments weren't addressed or only partially addressed. 

Im pretty sure after your changes this code will not run nor type check either. Make sure it does, instructions are in AGENTS.md.

Remove new Claude temp files.

If you have any questions about my comments I'm happy to answer, just tag me.

**Comment by ArkVex** [comment]
Nooo wayyy 😭 
I will take some time before submitting for review...sorry @aldenks i wont let claude messup the next pr 🙏 
Thankyou for your patience 😭 😭

**Comment by ArkVex** [comment]
@aldenks I've gone through all your feedback and made some changes so i thought I must consult you before commiting again...Here is the changes that i have done. Plz lemme know if I miss anything :)

**What I fixed:**
- Removed the `HasTimeInfo` Protocol
- Simplified `region_job.py` to just 2 lines - now passes `process_results` directly
- Removed the conditional check when updating coordinates (fail early like you mentioned)
- Changed function to return the dataset instead of None
- Replaced warning with assert
- Cleaned up all the unnecessary comments in tests
- Removed unused imports

**Tests added:**
- Test for unmodified init_times staying unchanged
- Integration test in `dynamical_dataset_test.py` checking ingested_forecast_length = 3h

**Everything passes:**
- ty check ✓
- ruff check ✓
- all unit tests ✓
- integration test ✓

**Comment by ArkVex** [comment]
I might have missed somethings again...plz lemme know your thoughts.
And yes I am open to criticism 😭

File: tmpclaude-0904-cwd [hunk 11]
```diff
@@ -0,0 +1 @@
+/c/Users/Lenovo/OneDrive/Desktop/reformatters
```
Comment (aldenks) [inline tmpclaude-0904-cwd]: Please remove all these tmpclaude files

File: src/reformatters/noaa/gfs/region_job.py [hunk 12]
```diff
@@ -150,6 +151,20 @@ def apply_data_transformations(
...
+    ) -> xr.Dataset:
+        ds = super().update_template_with_results(process_results)
+
+        all_coords = []
+        for coord_list in process_results.values():
+            all_coords.extend(coord_list)
+
+        update_ingested_forecast_length(ds, all_coords)
+
+        return ds
```
Comment (aldenks) [inline src/reformatters/noaa/gfs/region_job.py]: I think your changes here didn't get pushed?

File: src/reformatters/common/ingest_stats.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+log = get_logger(__name__)
+
+
+class DeterministicForecastSourceFileCoord(Protocol):
+    init_time: Timestamp
+    lead_time: Timedelta
+
+
+class HasTimeInfo(Protocol):
```
Comment (aldenks) [inline src/reformatters/common/ingest_stats.py]: Remove

File: tests/common/test_ingest_stats.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class MockSourceFileCoord(SourceFileCoord):
+    init_time: Timestamp
+    lead_time: Timedelta
+
+    def out_loc(self) -> Mapping[Dim, CoordinateValueOrRange]:
+        return {}
+
+
```
Comment (aldenks) [inline tests/common/test_ingest_stats.py:18]: We are missing a test that checks that the existing values in the array not *not* modified.

**Review by aldenks — CHANGES_REQUESTED** [review]
Hi @ArkVex, didn't mean to leave you hanging. I'll be a bit slow this week, I'm on vacation. 

You're much closer! I think some changes didn't get pushed?

I don't see the dynamical dataset integration test you mentioned

**Review by aldenks — COMMENTED** [review]
Hi @ArkVex, didn't meant to leave you hanging, I've been on vacation so I'll be slower to respond this week.

This is much closer! See individual comments above. 

I think some of your changes might not have been pushed. I don't see the integration test you mentioned. 

Also look at the failed "checks" logs each time you push if you aren't able to run them locally. Theres some type checking errors in there now

**Comment by ArkVex** [comment]
@aldenks I didnt push the code yet 😢 
I just wanted you to confirm the changes that i have mentioned above...
I am pushing the code now...plz take your time to review i am not in a hurry :)

File: tests/common/test_ingest_stats.py [hunk 15]
```diff
@@ -23,7 +22,7 @@ def test_update_ingested_forecast_length_simple() -> None:
         pd.Timestamp("2025-01-01 18:00"),
     ]
 
-    empty_deltas = np.array([pd.NaT, pd.NaT], dtype="timedelta64[ns]")
+    empty_deltas = pd.to_timedelta([pd.NaT, pd.NaT]).to_numpy()  # type: ignore[call-overload]
```
Comment (ArkVex) [inline tests/common/test_ingest_stats.py:25]: @aldenks just a headsup this line was failing in type so i had to ignore this line while type checking.

**Review by aldenks — COMMENTED** [review]
Looking great @ArkVex. Next steps:
- Fix test failures (see Checks logs)
- add test that checks that existing init times which are not processed (no source file coords for them) have their ingested forecast length coordinate values left unchanged 
- implement this for NOAA HRRR forecast 48 hour 

After that we could implement for the ensemble datasets (NOAA GEFS and ECMWF IFS ENS) but I'd recommend leaving that as a follow up PR because they have an additional ensemble member dimension.

**Comment by ArkVex** [comment]
Hey @aldenks @JackKelly its been a long time. I am sorry I got a lil occupied in some work. just wanted to confirm if this issue is still open to work?

**Comment by aldenks** [comment]
@ArkVex yes still good to work on if you're interested

**Comment by ArkVex** [comment]
Thankyou...I will get started from where I left 😄

**Comment by ArkVex** [comment]
@aldenks Ptal when you are available

**Comment by aldenks** [comment]
haven't forgot about this, will get to it soon

**Comment by ArkVex** [comment]
@aldenks Bro 😢

**Comment by ArkVex** [comment]
Ahhh..guys canl I get some review?
If not can I work on a different issue?
